client_plugin/drivers/shared/etcdops.go

```diff
@@ -0,0 +1,271 @@
+func joinEtcdCluster(nodeAddr string, leaderAddr string, nodeID string) error {
+	endpoint := leaderAddr + etcdClientPort
+	cfg := etcdClient.Config{
+		Endpoints: []string{endpoint},
+	}
+
+	cli, err := etcdClient.New(cfg)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"endpoint": endpoint,
+				"error": err},
+		).Error("Failed to create ETCD Client ")
+		return err
+	}
+	defer cli.Close()
+
+	// list all current ETCD members, check if this node is already added as a member
+	lresp, err := cli.MemberList(context.Background())
+	if err != nil {
+		log.WithFields(
+			log.Fields{"endpoint": endpoint,
+				"error":       err,
+				"members len": len(lresp.Members)},
+		).Error("Failed to list member for ETCD")
+		return err
+	}
+
+	peerAddr := etcdScheme + nodeAddr + etcdPeerPort
+	existing := false
+	for _, member := range lresp.Members {
+		if member.PeerURLs[0] == peerAddr {
+			if member.Name == "" {
+				// already added, not started
+				log.WithFields(
+					log.Fields{"nodeID": nodeID,
+						"peerAddr": peerAddr},
+				).Info("This node is already added as a etcd member, but not started yet ")
+
+				existing = true
+			} else {
+				// already added and started, need to reset
+				log.WithFields(
+					log.Fields{"nodeID": nodeID,
+						"peerAddr": peerAddr},
+				).Info("This node is already added as a etcd member and started, need to remove first ")
+
+				_, err = cli.MemberRemove(context.Background(), member.ID)
+				if err != nil {
+					log.WithFields(
+						log.Fields{"peerAddr": peerAddr,
+							"member.ID": member.ID},
+					).Error("Failed to remove member for ETCD")
+					return err
+				}
+			}
+			break
+		}
+	}
+
+	initCluster := ""
+	if !existing {
+		peerAddrs := []string{peerAddr}
+		aresp, err := cli.MemberAdd(context.Background(), peerAddrs)
+		if err != nil {
+			log.WithFields(
+				log.Fields{"endpoint": endpoint,
+					"error":       err,
+					"members len": len(aresp.Members)},
+			).Error("Failed to add member for ETCD")
+			return err
+		}
+		for _, member := range aresp.Members {
+			if member.Name != "" {
+				initCluster += member.Name + "=" + member.PeerURLs[0] + ","
+			}
+		}
+	} else {
+		for _, member := range lresp.Members {
+			if member.Name != "" {
+				initCluster += member.Name + "=" + member.PeerURLs[0] + ","
+			}
+		}
+	}
+
+	lines := []string{
+		"--name", nodeID,
+		"--advertise-client-urls", etcdScheme + nodeAddr + etcdClientPort,
+		"--initial-advertise-peer-urls", etcdScheme + nodeAddr + etcdPeerPort,
+		"--listen-client-urls", etcdScheme + etcdListenURL + etcdClientPort,
+		"--listen-peer-urls", etcdScheme + etcdListenURL + etcdPeerPort,
+		"--initial-cluster-token", etcdClusterToken,
+		"--initial-cluster", initCluster + nodeID + "=" + etcdScheme + nodeAddr + etcdPeerPort,
+		"--initial-cluster-state", etcdClusterStateExisting,
+	}
+
+	go etcdRoutine(lines)
+
+	return nil
+}
+
+func etcdRoutine(cmd []string) {
```

Comment:
the name feels wrong. etcdService() or etcdLoop() or even etcdDaemon() are closer to what's really going on